File: sdks/python/src/ankka/context.py
"""What a handler can see about the call it is handling."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING, Any

from ankka._proto.ankka.protocol.v1 import payload_pb2

if TYPE_CHECKING:
    from ankka.client import ComponentClient


@dataclass(frozen=True)
class Metadata:
    """Key/value pairs that travel with a call. Opaque to a handler: the sidecar puts the trace
    and span ids in here, and copying it onto a nested call is what makes that call a child span."""

    entries: tuple[tuple[str, str], ...] = ()

    def get(self, key: str) -> str | None:
        for k, v in self.entries:
            if k.lower() == key.lower():
                return v
        return None

    def get_all(self, key: str) -> list[str]:
        return [v for k, v in self.entries if k.lower() == key.lower()]

    def set(self, key: str, value: str) -> Metadata:
        rest = tuple((k, v) for k, v in self.entries if k.lower() != key.lower())
        return Metadata(rest + ((key, value),))

    def add(self, key: str, value: str) -> Metadata:
        return Metadata(self.entries + ((key, value),))

    def to_pb(self) -> payload_pb2.Metadata:
        return payload_pb2.Metadata(entries=[payload_pb2.Metadata.Entry(key=k, value=v) for k, v in self.entries])

    @staticmethod
    def from_pb(pb: payload_pb2.Metadata | None) -> Metadata:
        if pb is None:
            return Metadata()
        return Metadata(tuple((e.key, e.value) for e in pb.entries))


EMPTY = Metadata()


@dataclass(frozen=True)
class CommandContext:
    """What an entity or workflow handler sees: which instance, which component, the call's
    metadata, the sequence number, and the client for calling other components."""

    entity_id: str
    component_id: str
    metadata: Metadata
    sequence_number: int
    client: ComponentClient


@dataclass(frozen=True)
class Principal:
    subject: str
    name: str | None = None
    email: str | None = None
    email_verified: bool = False
    roles: frozenset[str] = frozenset()


@dataclass(frozen=True)
class RequestContext:
    """What an endpoint handler sees beyond its typed arguments: query parameters (repeatable),
    headers, the principal when the ACL established one, and the request's trace metadata."""

    query: tuple[tuple[str, str], ...] = ()
    headers: tuple[tuple[str, str], ...] = ()
    principal: Principal | None = None
    metadata: Metadata = field(default_factory=Metadata)

    def query_param(self, name: str) -> str | None:
        for k, v in self.query:
            if k == name:
                return v
        return None

    def query_params(self, name: str) -> list[str]:
        return [v for k, v in self.query if k == name]

    def header(self, name: str) -> str | None:
        for k, v in self.headers:
            if k.lower() == name.lower():
                return v
        return None


@dataclass(frozen=True)
class SessionContext:
    """What an agent handler or tool sees."""

    session_id: str
    component_id: str
    metadata: Metadata
    client: ComponentClient


def _unused(*_: Any) -> None:  # keeps the TYPE_CHECKING import honest for linters
    return None

File: sdks/typescript/src/context.ts
// What a handler can see about the request it is running for. A `CommandContext` is bound on the
// component instance for one call; an HTTP `RequestContext` lives in an `AsyncLocalStorage`, the Node
// equivalent of the thread-local the Scala SDK uses and the `ContextVar` the Python SDK uses: one
// request per asynchronous context, and work handed to another context cannot see it.

import { AsyncLocalStorage } from "node:async_hooks"
import { Instant } from "./time.ts"
import type { Metadata } from "./effects/common.ts"
import type { Metadata as ProtoMetadata } from "./_proto/ankka/protocol/v1/payload_pb.ts"

/** The context of one command on a stateful component. */
export interface CommandContext {
  readonly componentId: string
  readonly entityId: string
  /** The sequence number of the last event applied before this command; `0n` for a fresh entity. */
  readonly sequenceNumber: bigint
  readonly metadata: Metadata
  /** The current instant; a seam a test can replace. */
  now(): Instant
}

export function commandContext(componentId: string, entityId: string, sequenceNumber: bigint, metadata: Metadata, now: () => Instant = Instant.now): CommandContext {
  return Object.freeze({ componentId, entityId, sequenceNumber, metadata, now })
}

export function metadataFromProto(proto: ProtoMetadata | undefined): Metadata {
  const out: Record<string, string> = {}
  for (const e of proto?.entries ?? []) out[e.key] = e.value
  return Object.freeze(out)
}

export function metadataToProto(metadata: Metadata | undefined): { entries: { key: string; value: string }[] } {
  return { entries: Object.entries(metadata ?? {}).map(([key, value]) => ({ key, value })) }
}

/** The caller an authenticated route was invoked by. */
export interface Principal {
  readonly subject: string
  readonly name: string | null
  readonly email: string | null
  readonly emailVerified: boolean
  readonly roles: readonly string[]
}

/** Query parameters in request order; a name may repeat. */
export class Query {
  private readonly pairs: readonly (readonly [string, string])[]

  constructor(pairs: readonly (readonly [string, string])[]) {
    this.pairs = pairs
    Object.freeze(this)
  }
  get(name: string): string | null {
    const hit = this.pairs.find(([n]) => n === name)
    return hit ? hit[1] : null
  }
  getAll(name: string): string[] {
    return this.pairs.filter(([n]) => n === name).map(([, v]) => v)
  }
  has(name: string): boolean {
    return this.pairs.some(([n]) => n === name)
  }
  entries(): readonly (readonly [string, string])[] {
    return this.pairs
  }
}

/** Request headers; names are matched case-insensitively. */
export class Headers {
  private readonly pairs: readonly (readonly [string, string])[]

  constructor(pairs: readonly (readonly [string, string])[]) {
    this.pairs = pairs
    Object.freeze(this)
  }
  get(name: string): string | null {
    const lower = name.toLowerCase()
    const hit = this.pairs.find(([n]) => n.toLowerCase() === lower)
    return hit ? hit[1] : null
  }
  getAll(name: string): string[] {
    const lower = name.toLowerCase()
    return this.pairs.filter(([n]) => n.toLowerCase() === lower).map(([, v]) => v)
  }
  entries(): readonly (readonly [string, string])[] {
    return this.pairs
  }
}

/** What an endpoint handler sees of its request: `this.request` on the endpoint. `P` is the typed path parameters. */
export interface RequestContext<P = Readonly<Record<string, unknown>>> {
  readonly params: P
  readonly query: Query
  readonly headers: Headers
  /** The caller, when the route's access rule is `authenticated`; `null` otherwise. */
  readonly principal: Principal | null
  readonly metadata: Metadata
}

const requestStorage = new AsyncLocalStorage<RequestContext>()

/** Runs `fn` with `request` as the current request, for the asynchronous work it starts. */
export function withRequest<T>(request: RequestContext, fn: () => T): T {
  return requestStorage.run(request, fn)
}

/** The current request, or `undefined` outside one. */
export function requestIfAny(): RequestContext | undefined {
  return requestStorage.getStore()
}

/** The current request, or a clear error when called outside one. */
export function currentRequest(): RequestContext {
  const r = requestStorage.getStore()
  if (!r) throw new Error("request is only available inside an endpoint handler")
  return r
}
